File: python/maxAreaOfIslands.py
from typing import List

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid or not grid[0]:
            return 0

        length = len(grid)
        width = len(grid[0])
        visited = [[False] * width for _ in range(length)]
        max_area = 0

        def dfs(i, j):
            if i < 0 or i >= length or j < 0 or j >= width or visited[i][j] or grid[i][j] == 0:
                return 0
            visited[i][j] = True
            return (1 + dfs(i, j-1) + dfs(i, j+1) + dfs(i-1, j) + dfs(i+1, j))

        for i in range(length):
            for j in range(width):
                if not visited[i][j] and grid[i][j] == 1:
                    max_area = max(max_area, dfs(i, j))

        return max_area
        

class Solution2:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visit = set()

        def dfs(r, c):
            if (
                r < 0
                or r == ROWS
                or c < 0
                or c == COLS
                or grid[r][c] == 0
                or (r, c) in visit
            ):
                return 0
            visit.add((r, c))
            return 1 + dfs(r + 1, c) + dfs(r - 1, c) + dfs(r, c + 1) + dfs(r, c - 1)

        area = 0
        for r in range(ROWS):
            for c in range(COLS):
                area = max(area, dfs(r, c))
        return area

File: TestLeetCode/Graph/MaxAreaOfIslands.cs
public class MaxAreaOfIslandSolution {
    public int MaxAreaOfIsland(int[][] grid) {
        if (grid == null || grid.Length == 0) { return 0; }
        if (grid[0].Length == 0) { return 0; }

        int length = grid.Length;
        int width = grid[0].Length;

        var visited = new bool[length][];
        for (int i = 0; i < length; i++) {
            visited[i] = new bool[width];
        }

        int max = 0;
        int cur = 0;

        for (int i = 0; i < length; i++) {
            for (int j = 0; j < width; j++) {
                if (!visited[i][j] && grid[i][j] == 1) {
                    cur = dfs(grid, visited, i, j);
                    max = Math.Max(max, cur);
                }
            }        
        }

        return max;
    }

    private int dfs(int[][] grid, bool[][] visited, int i, int j) {
        if (i < 0 || i >= grid.Length || j < 0 || j >= grid[0].Length ||
            visited[i][j] || grid[i][j] == 0) { return 0; }
        visited[i][j] = true;

        return 1 + dfs(grid, visited, i, j-1) + dfs(grid, visited, i, j+1)
            + dfs(grid, visited, i-1, j) + dfs(grid, visited, i+1, j);
    }
}
